### PR TITLE
feat(presets): Preset that applies a label for the different update types

### PR DESCRIPTION
**Pull request by astellingwerf**
## Changes:

There's a new preset that generates a package rule per allowed update Type, that applies a label. The label is sandwiched between the two arguments.

## Context:

I've build a piece of code in my own configuration, but I imagine that this can be beneficial to many users, not just me. By setting it up like this, you're also guaranteed to cover any new update type that may arise.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/presets/internal/default.ts [hunk 1]
```diff
@@ -456,6 +459,18 @@ export const presets: Record<string, Preset> = {
...
+    description:
+      'Apply label per update type, prefixed by <code>{{arg0}}</code> and suffixed by <code>{{arg1}}</code> to PRs',
+    packageRules: [].concat(
+      options
+        .find((it) => it.name === 'matchUpdateTypes')
+        .allowedValues.map((t) => ({
+          matchUpdateTypes: [t],
+          labels: ['{{arg0}}' + t + '{{arg1}}'],
+        }))
+    ),
```
Comment (viceice) [inline lib/config/presets/internal/default.ts:472]:
```suggestion
    packageRules:
      options
        .find((it) => it.name === 'matchUpdateTypes')
        .allowedValues.map((t) => ({
          matchUpdateTypes: [t],
          labels: ['{{arg0}}' + t + '{{arg1}}'],
        })),
```

The `concat` isn't nesessary, as `map` is already creating a new array.

Comment (astellingwerf) [reply lib/config/presets/internal/default.ts:472]: Let me try that again: I did it that way first, but couldn't get it to compile, and this was a quick work around.

Comment (astellingwerf) [reply lib/config/presets/internal/default.ts]: Addressed it with an `as PackageRule`.

File: lib/config/presets/internal/default.ts [hunk 2]
```diff
@@ -456,6 +460,19 @@ export const presets: Record<string, Preset> = {
...
+    description:
+      'Apply label per update type, prefixed by <code>{{arg0}}</code> and suffixed by <code>{{arg1}}</code> to PRs',
+    packageRules: options
+      .find((it) => it.name === 'matchUpdateTypes')
+      .allowedValues.map(
+        (t) =>
+          ({
+            matchUpdateTypes: [t],
+            labels: ['{{arg0}}' + t + '{{arg1}}'],
+          } as PackageRule)
```
Comment (viceice) [inline lib/config/presets/internal/default.ts:473]:
```suggestion
          })
```

File: lib/config/presets/internal/default.ts [hunk 3]
```diff
@@ -456,6 +460,19 @@ export const presets: Record<string, Preset> = {
       'Apply labels <code>{{arg0}}</code> and <code>{{arg1}}</code> to PRs',
     labels: ['{{arg0}}', '{{arg1}}'],
   },
+  labelUpdateTypes: {
+    description:
+      'Apply label per update type, prefixed by <code>{{arg0}}</code> and suffixed by <code>{{arg1}}</code> to PRs',
+    packageRules: options
+      .find((it) => it.name === 'matchUpdateTypes')
+      .allowedValues.map(
+        (t) =>
```
Comment (viceice) [inline lib/config/presets/internal/default.ts:469]:
```suggestion
        (t: UpdateType) =>
```

**Review by viceice — COMMENTED** [review]
I think this is better

File: lib/config/presets/internal/default.ts (hunk 3 above)
Comment (astellingwerf) [reply lib/config/presets/internal/default.ts]: That doesn't work, as the `.allowedValues` is of type `string[]`. I can map to `UpdateType` prior to mapping to a `PackageRule`.

**Comment by rarkins** [comment]
Sorry, what's the use case here? I don't understand the significant if Eg pre-pin-post

**Comment by astellingwerf** [comment]
It means that you can have labels not just be the name of the update type, but also have them like _renovate-update-major_

**Comment by astellingwerf** [comment]
@rarkins how can I help progressing this PR?

File: lib/config/presets/internal/default.ts [hunk 4]
```diff
@@ -456,6 +460,17 @@ export const presets: Record<string, Preset> = {
...
   },
+  labelUpdateTypes: {
+    description:
+      'Apply label per update type, prefixed by <code>{{arg0}}</code> and suffixed by <code>{{arg1}}</code> to PRs',
+    packageRules: options
+      .find((it) => it.name === 'matchUpdateTypes')
+      .allowedValues.map((t) => t as UpdateType)
+      .map((t) => ({
+        matchUpdateTypes: [t],
+        labels: ['{{arg0}}' + t + '{{arg1}}'],
```
Comment (viceice) [inline lib/config/presets/internal/default.ts:471]: This will override any previous labels.

**Review by viceice — COMMENTED** [review]
I don't see a big use case for others. This looks like a very specific usecase. so i would decline this.

@rarkins @JamieMagee Do you agree?

**Comment by rarkins** [comment]
> @rarkins how can I help progressing this PR?

I think we need to find a way to document this, especially using 1-2 examples. But I'm not sure where.

File: lib/config/presets/internal/default.ts (hunk 4 above)
Comment (astellingwerf) [reply lib/config/presets/internal/default.ts:471]: How is this different from other presets? I'm creating the same objects as its siblings, right?

**Comment by astellingwerf** [comment]
After browsing the documentation, I came across writing on Template fields. If it works like this, maybe I should retract this PR:
```
addLabels: ['renovate-update-{{updateType}}']
```

**Comment by viceice** [comment]
> After browsing the documentation, I came across writing on Template fields. If it works like this, maybe I should retract this PR:
> 
> ```
> addLabels: ['renovate-update-{{updateType}}']
> ```

Yes, `addLabels` is already templatable, so i think we don't need this preset.

File: lib/config/presets/internal/default.ts (hunk 4 above)
Comment (astellingwerf) [reply lib/config/presets/internal/default.ts:471]: @viceice Shouldn't we fix all current templates from using `labels` to `addLabels`?

Comment (viceice) [reply lib/config/presets/internal/default.ts:471]: 🤷‍♂️ @rarkins WDYT?

**Pull request closed without merging**